ENH: adding autofilter when writing to excel (pandas-dev#61194)

jarent-nvidia opened the pull request:
- [x] closes #61194 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst` file if fixing a bug or adding a new feature.

Description of the change:
Added boolean 'autofilter' optional argument to 'DataFrame.to_excel'  function to add automatic filters to all columns in the exported excel files. As discussed with @WillAyd starting simple by allowing to add filters only to all or none columns. 
New option supported by 'xlsxwriter' and 'openpyxl' engines.

jarent-nvidia committed:
ENH: adding autofilter when writing to excel (pandas-dev#61194)

jarent-nvidia committed:
chore: Added missing documentation

jarent-nvidia committed:
chore: Fixed documentation

WillAyd commented on pandas/io/formats/excel.py:
```diff
@@ -873,6 +877,16 @@ def get_formatted_cells(self) -> Iterable[ExcelCell]:
+        Convert 0-based column index to Excel column name.
+        """
+        column_name = ""
+        while index > 0 or not column_name:
```
Can you raise when index < 0? Also, what's the point of this being a `while` loop?

jarent-nvidia replied:
@WillAyd Thanks for the feedback - input validation added.

WillAyd reviewed (changes requested):
Small comment but overall looks like a really nice PR. @rhshadrach any thoughts?

jarent-nvidia committed:
chore: Aded input validation to ensue that column index  > 0

jarent-nvidia committed:
fix: Remove print statement

WillAyd reviewed (approved):
lgtm - let's see what others think

rhshadrach commented on pandas/io/excel/_odswriter.py:
```diff
@@ -99,6 +99,7 @@ def _write_cells(
         startrow: int = 0,
         startcol: int = 0,
         freeze_panes: tuple[int, int] | None = None,
+        autofilter_range: str | None = None,
```
It seems to me we should be raising an error here if a user specifies `autofilter=True` and it is not supported with an engine rather than silently ignore the user's specification.

rhshadrach reviewed (changes requested):
Agreed this is looking very good. Can you add tests for:

- `index=True`
- nonzero startrow / startcol
- A DataFrame with MultiIndex columns (also called a hierarchical columns) with both `merge_cells=True` and `merge_cells=False`.

jarent-nvidia committed:
fix: Ensure that ndex size is considered when setting autofilter range.
feat: Throw an error when autofilter set for 'odf' engine.
feat: Shows warning when autofilter is set together with merge_cells.
chore: Added tests for index=True, startrow/startcol, multindex with merge_cells.

jarent-nvidia committed:
chore: Check if numpy unit test failure is intermittent.

jarent-nvidia committed:
Merge branch 'main' into main

jarent-nvidia commented:
@rhshadrach I believe the issues I got with the checks is not related to my changes. I'm waiting for https://github.com/pandas-dev/pandas/pull/63107 to get all the checks green. Besides that the code changes are ready for your review. I hope I was able to address your feedback. List of changes:

fix: Ensure that index size is considered when setting autofilter range.
feat: Throw an error when autofilter set for 'odf' engine.
feat: Show warning when autofilter is set together with merge_cells.
chore: Added tests for index=True, startrow/startcol, multindex with merge_cell

jarent-nvidia committed:
Merge branch 'main' into main

rhshadrach reviewed (changes requested):
Previous request was for a DataFrame with MultiIndex _columns_:

```python
columns = MultiIndex(
    levels=[["x", "y"], ["w", "t"]],
    codes=[[0, 0, 1], [0, 1, 0]],
)
df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=columns)
```

but agreed as you've found a MultiIndex index also gives problems. I think we should raise when either `columns` or `index` is a MultiIndex, `autofilter=True`, and `merge_cells=True`.

jarent-nvidia committed:
chore: move autofilter tests from test_style to test_writers
fix: Handle multindex columns offset
feat: If autofilter is set together with merge_cells, raise an exception

rhshadrach commented on pandas/io/formats/excel.py:
```diff
@@ -532,6 +532,8 @@ class ExcelFormatter:
         It should have signature css_declarations string -> excel style.
         This is only called for body cells.
+    autofilter : bool, default False
+        If True, add automatic filters to all columns
```
nit

```suggestion
        If True, add automatic filters to all columns.
```

jarent-nvidia replied:
Done!

rhshadrach reviewed (changes requested):
Looks great! One last nit and we're good here.

jarent-nvidia committed:
chore: Fix docstrings

rhshadrach reviewed (approved):
lgtm
